Add command-line automation features

lkrms opened the pull request:
Issue: https://github.com/hovancik/stretchly/issues/458

### Requirements

- [x]  issue was opened to discuss proposed changes before starting implementation.
- [ ]  during development, `node` version specified in `package.json` was used (ie using [nvm](https://github.com/creationix/nvm)).
- [x]  package versions and package-lock.json were not changed (`npm install --no-save`).
- [x]  app version number was not changed.
- [ ]  all new code has tests to ensure against regressions.
- [x] `npm run lint` reports no offenses.
- [ ] `npm run test` is error-free.
- [ ]  README and CHANGELOG were updated accordingly.

### Description of the Change

Commenced implementation of changes discussed in https://github.com/hovancik/stretchly/issues/458

For now, the following is supported:

```
Usage: stretchly <command> [arguments]

  Commands available:
    reset-breaks
    skip-to-microbreak
    skip-to-break
    pause-breaks [indefinitely|until-morning|<seconds>]
      Default: indefinitely
    resume-breaks
```

### Verification Process

I've confirmed that the commands listed above are received and actioned correctly by the running instance of Stretchly, whether called via `npm start` or from a freshly built AppImage.

### Other information

I anticipate additional commands and improvements will be needed before this is production-ready.

SeabassHeyneman commented:
I checked out the branch and ran it on my end. It works for me!

hovancik commented:
Thanks, trying now but I'll take a deeper look. There are some things to consider.

lkrms commented:
@SeabassHeyneman glad it's working for you. It would be great if you could test the improvements I just added too :)

@hovancik no worries. I've fixed some basic stuff (e.g. which instance outputs usage information) and added the `set-next-idea` command put forward in the original feature request. It made sense to add optional next-idea arguments to the `skip-to-next-*` commands at the same time. The current usage output looks like this:

```
stretchly is already running.

To send a command to the running instance of stretchly:
  stretchly <command> [arguments]

Commands available:
  reset-breaks
  skip-to-microbreak [<idea>]
  skip-to-break [<idea> [<idea-text>]]
  pause-breaks [indefinitely|until-morning|<seconds>]
    Default: indefinitely
  resume-breaks
  set-next-idea <idea> [<idea-text>]
    If the next break is a microbreak, <idea-text> will be ignored.

Examples:
  stretchly pause-breaks until-morning
  stretchly skip-to-break "Remember" "It's time for lunch"
  stretchly set-next-idea "1. Take a breath. 2. Start on $TASK."
```

hovancik commented:
Thanks @lkrms ;] 

I am looking at tools like https://github.com/tj/commander.js and wonder what's the usual way to do args, whether it should be `stretchly pause-breaks until-morning` or `stretchly --pause-breaks=until-morning`.  Not sure how it then would be with `skip-to-break` and multiple params. What do you think? I am not really using much command line tools. 

Another thing: I am working on whole big redesign in `redesign` branch and it will be next release. Over there there are `Mini Break` and `Long Break` instead of `Microbreak` and `Break`. Do you wanna rebase and PR to there? There are many many changes there, including calling the app with capital s `Stretchly `

SeabassHeyneman commented:
👋 @lkrms, anything I can do to be helpful here? Let me know! 😃

hovancik commented:
@SeabassHeyneman what interface would you prefer? `stretchly pause-breaks until-morning` or `stretchly --pause-breaks=until-morning`

hovancik commented:
I am leaning toward ones with `--`, seems to me that they are `"standard"`

SeabassHeyneman commented:
I agree with you @hovancik , I prefer the ones with `--` because they are more standard.

SeabassHeyneman commented:
We're so close to the finish line. Can I help get us across?

hovancik commented:
Hi @SeabassHeyneman , I am having job interviews so a bit busy, but this and issue with linux fullscreen are next things on my todo ;]

SeabassHeyneman commented:
Amazing!

hovancik commented:
Actually, looking more into it, seems to me it should be like this: 
```
Usage: stretchly <command> [options]

Commands:
  stretchly mini [options]   Skip to next Mini Break
  stretchly long [options]   Skip to next Long Break
  stretchly pause [options]  Pause breaks
  stretchly resume           Resume breaks
  stretchly reset            Reset breaks

Options:
  --version       Show version number                                  [boolean]
  -t, --text      Specify text for next break idea (for Long Break)
  -d, --duration  Specify duration for pausing breaks in seconds
  -h, --help      Show help                                            [boolean]

Examples:
  stretchly pause                           Pauses breaks indefinitely
  stretchly pause -d 3600                   Pauses breaks for one hour
  stretchly resume                          Resumes breaks
  stretchly reset                           Resets breaks
  stretchly mini                            Skips to next Mini Break
  stretchly mini -T "Stretch up!"           Skips to next Mini Break with
                                            "Stretch up!" title
  stretchly -T "Stretch up!"                Sets title for next break idea to
                                            "Stretch up!"
  stretchly -T "Stretch up!" -t "Go         Sets title for next break idea to
  stretch, for real!!!"                     "Stretch up!" and body to "Go
                                            stretch, for real!!!"

```

Does it makes sense? 

There is a difference between `command` and `options`.

SeabassHeyneman commented:
This looks great

On Wed, May 27, 2020 at 6:20 AM Jan Hovancik <notifications@github.com>
wrote:

> Actually, looking more into it, seems to me it should be like this:
>
> Usage: stretchly <command> [options]
>
> Commands:
>   stretchly mini [options]   Skip to next Mini Break
>   stretchly long [options]   Skip to next Long Break
>   stretchly pause [options]  Pause breaks
>   stretchly resume           Resume breaks
>   stretchly reset            Reset breaks
>
> Options:
>   --version       Show version number                                  [boolean]
>   -t, --text      Specify text for next break idea (for Long Break)
>   -d, --duration  Specify duration for pausing breaks in seconds
>   -h, --help      Show help                                            [boolean]
>
> Examples:
>   stretchly pause                           Pauses breaks indefinitely
>   stretchly pause -d 3600                   Pauses breaks for one hour
>   stretchly resume                          Resumes breaks
>   stretchly reset                           Resets breaks
>   stretchly mini                            Skips to next Mini Break
>   stretchly mini -T "Stretch up!"           Skips to next Mini Break with
>                                             "Stretch up!" title
>   stretchly -T "Stretch up!"                Sets title for next break idea to
>                                             "Stretch up!"
>   stretchly -T "Stretch up!" -t "Go         Sets title for next break idea to
>   stretch, for real!!!"                     "Stretch up!" and body to "Go
>                                             stretch, for real!!!"
>
>
> Does it makes sense?
>
> There is a difference between command and options.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/hovancik/stretchly/pull/567#issuecomment-634654738>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMT4VX52DTYADIJZJV726PDRTUHRLANCNFSM4MPX7C2A>
> .
>


-- 
Sebastian Heyneman
*Pronouns: he/him/his*
Data Scientist

Asana
<http://asana.com/?utm_source=Asana&utm_medium=email&utm_campaign=signature>
1550 Bryant Street, Suite 800
San Francisco, CA 94103
LinkedIn <https://www.linkedin.com/in/sebastianheyneman/> | 415-513-7629

lkrms commented:
Sorry for going AWOL on you.

I had been thinking along the lines of "<binary_name> <command> <--option...>", as adopted by `npm`, `pip`, `composer`, etc., so I'm on board with this, @hovancik 

Happy to do the code re-work to adopt it, too. My preference would be to avoid pulling in a new dependency just for simple command-line argument processing, but maybe that's just me. I'd also suggest having a "long" version of each option name.

SeabassHeyneman commented:
Awesome, love the idea to have a "long" version of each option name .

On Wed, May 27, 2020 at 7:39 PM Luke Arms <notifications@github.com> wrote:

> Sorry for going AWOL on you.
>
> I had been thinking along the lines of "<binary_name> <--option...>", as
> adopted by npm, pip, composer, etc., so I'm on board with this, @hovancik
> <https://github.com/hovancik>
>
> Happy to do the code re-work to adopt it, too. My preference would be to
> avoid pulling in a new dependency just for simple command-line argument
> processing, but maybe that's just me. I'd also suggest having a "long"
> version of each option name.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/hovancik/stretchly/pull/567#issuecomment-635060020>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMT4VXYB5ABPXFZAWHGQCSTRTXFHDANCNFSM4MPX7C2A>
> .
>


-- 
Sebastian Heyneman
*Pronouns: he/him/his*
Data Scientist

Asana
<http://asana.com/?utm_source=Asana&utm_medium=email&utm_campaign=signature>
1550 Bryant Street, Suite 800
San Francisco, CA 94103
LinkedIn <https://www.linkedin.com/in/sebastianheyneman/> | 415-513-7629

hovancik commented:
@lkrms if you'd reworked it, that would be awesome :) I could go work on other things :+1:  

I've been trying some libraries but only had problems ;/

some thoughts: 
- would be best to start from `redesign` branch, there are big changes as I mentioned
- I've seen some patter then I like, pasting here

```
const gotTheLock = app.requestSingleInstanceLock()

if (!gotTheLock) {
 // code for seconds instance 
  app.quit()
  return
} else {
  app.on('second-instance', (event, args, workingDirectory) => {
    // code for running instance
  }
}
```

- Also maybe not clear from my help example earlier, but my thought was to: 
`-T`, `--title`: title in Long Break and Mini Break
`-t`, `--text`: text for Long Break (as Mini Break has only title)

I also had other idea, based on for example `heroku/dokku`: 
- `breaks:pause [indefinitely|until-morning|seconds]` - defaults to indef.
- `breaks:resume`
- `breaks:reset`
- `breaks:start:mini [default|title]` - defaults from app
- `breaks:start:long [default|title] [default|text]`
- `breaks:idea:mini [title]` - set idea for next mini break
- `breaks:idea:long [title] [text]` - set idea for next long break
This would allow better way to set idea for breaks

SeabassHeyneman commented:
I really like the format you're citing from `heroku/dokku`

On Thu, May 28, 2020 at 1:57 AM Jan Hovancik <notifications@github.com>
wrote:

> @lkrms <https://github.com/lkrms> if you'd reworked it, that would be
> awesome :) I could go work on other things 👍
>
> I've been trying some libraries but only had problems ;/
>
> some thoughts:
>
>    - would be best to start from redesign branch, there are big changes
>    as I mentioned
>    - I've seen some patter then I like, pasting here
>
> const gotTheLock = app.requestSingleInstanceLock()
>
>
>
> if (!gotTheLock) {
>
>  // code for seconds instance
>
>   app.quit()
>
>   return
>
> } else {
>
>   app.on('second-instance', (event, args, workingDirectory) => {
>
>     // code for running instance
>
>   }
>
> }
>
>
>
>    - Also maybe not clear from my help example earlier, but my thought
>    was to:
>    -T, --title: title in Long Break and Mini Break
>    -t, --text: text for Long Break (as Mini Break has only title)
>
> I also had other idea, based on for example heroku/dokku:
>
>    - breaks:pause [indefinitely|until-morning|seconds] - defaults to
>    indef.
>    - breaks:resume
>    - breaks:reset
>    - breaks:start:mini [default|title] - defaults from app
>    - breaks:start:long [default|title] [default|text]
>    - breaks:idea:mini [title] - set idea for next mini break
>    - breaks:idea:long [title] [text] - set idea for next long break
>    This would allow better way to set idea for breaks
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/hovancik/stretchly/pull/567#issuecomment-635212811>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMT4VX3G7FJ7VMIKDQMOZQDRTYRPFANCNFSM4MPX7C2A>
> .
>


-- 
Sebastian Heyneman
*Pronouns: he/him/his*
Data Scientist

Asana
<http://asana.com/?utm_source=Asana&utm_medium=email&utm_campaign=signature>
1550 Bryant Street, Suite 800
San Francisco, CA 94103
LinkedIn <https://www.linkedin.com/in/sebastianheyneman/> | 415-513-7629

SeabassHeyneman commented:
Hi guys! Want to be clear, is @ikrms going to own next steps?

Thanks, and have a great weekend!

hovancik commented:
Well, I am counting on @lkrms  ;]

SeabassHeyneman commented:
Hi @lkrms! Are you still interested in completing this projects?

Let us know! 😀

SeabassHeyneman commented:
@hovancik wondering whether we need to get this over the finish line

hovancik commented:
- rebase with `redesign` and use it's language :scream: 
- design the interface (possibly one of the options we talked about) :speech_balloon: 
- code :woman_technologist: 
- merge :1st_place_medal:

SeabassHeyneman commented:
What would you feel comfortable taking on?

hovancik commented:
I did a rebase to redesign just now and pushed it to `automation-rebased-redesign` branch.

SeabassHeyneman commented:
Awesome!! So does that mean this will be available when you launch the redesign?

SeabassHeyneman commented:
OK, I found it. I'm going to download Stretchly with that branch and give it a try.

I'll share feedback with you, if that's alright. Thanks!!

hovancik commented:
It's only the first step, it has previous code without changes that are needed.

SeabassHeyneman commented:
No worries! I'll circle back in August

papey commented:
I'm super excited for this feature, tell me if you need help on something :+1:

hovancik commented:
Hey @papey , if you're willing, you could work on code :) 

This PR is very outdated, but I did rebase with newer code in branch `automation-rebased-redesign` some time ago. 

What needs to be done: 
- I guess rebase/merge my branch again with master
- change the names from Micro to Mini Break and from break to Long Break
- design the interface (possibly one of the options we talked about in previous comments)
- code it and test :)

papey commented:
Sorry for the delay, will try to work on it ASAP, since I can't contribute directly to this PR i will create a new one adding a comment to this one to point to the current active PR.

hovancik commented:
@papey you should start from branch `automation-rebased-redesign` - I did rebase there some time ago, so it's "newer" 

what interface (style of commands) would you like to do?

papey commented:
If you prefer the `heroku` style, i'll go with it

hovancik commented:
Well, TBH, not really using that much command line tools, so I'll leave it with you

papey commented:
Work started on PR https://github.com/hovancik/stretchly/issues/458

hovancik commented:
I'll close this one, then.
